### ISSUE TITLE
Room - #announce room can be created via room mention

### ISSUE THREAD

**Issue by IuliiaHerets** [Expensify/App#50173]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.44-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**Email or phone of affected tester (no customers):** applausetester+pso@applause.expensifail.com
**Issue reported by:** Applause Internal Team

## Action Performed:
1. Go to staging.new.expensify.com
2. Create a new workspace.
3. Go to FAB > Start chat > Room.
4. Enter #announce as room name and save it.
5. Note that it shows error that the room name cannot be #announce.
6. Go to workspace chat.
7. Send message containing #announce mention.
8. Click Yes from the whisper to create the room.
9. Note that #announce room can be created via mention.
10. Invite 3 members to the workspace.
11. Note that now there are two #announce rooms.

## Expected Result:
In Step 8, app should prevent #announce room from being created via room mention.

## Actual Result:
In Step 8. #announce room can be via room mention.
In Step 11, after inviting 3 members to the workspace, it results in two #announce rooms.

## Workaround:
Unknown

## Platforms:

- [x] Android: Native
- [x] Android: mWeb Chrome
- [x] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

https://github.com/user-attachments/assets/c0ea7054-272e-4018-a84a-b084b33dd9aa

  
</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by IuliiaHerets** [comment]
@sonialiap  FYI I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors

**Comment by mkzie2** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
In Step 8. #announce room can be via room mention.
In Step 11, after inviting 3 members to the workspace, it results in two #announce rooms.

### What is the root cause of that problem?
- The ```policyAnnounce``` is just created if the policy has 3 or more members. This is handled in [PR](https://github.com/Expensify/App/pull/48660).

- In this bug, when ```policyAnnounce``` is not created, then we send a message with room mention ```#announce```, BE returns the actionable whisper, that is incorrect.
### What changes do you think we should make in order to solve the problem?
- First, BE should not return actionable whisper message when user send ```#announce``` message.

- Then, we need to modify the error message when creating a new room as well as when editing a room name. When user try creating a room with "announce" as name:

1. If the ```policyAnnounce``` room has not been created yet, should show _"This name is reserved for future use. Please choose a different name"_

2. If the ```policyAnnounce``` room has been, should show _"A room with this name already exists"_ or _"Announce is a default room on all workspaces. Please choose another name."_

- Below is the code changes:


1. Introduce a function to check whether the policy has ```policyAnnounce``` or not:

```
function hasAnnounceRoom(reports: OnyxCollection<Report>, policyID: string): boolean {
    return Object.values(reports ?? {}).some((report) => report && report.policyID === policyID && isAnnounceRoom(report));
}
```


2. Update the validation:

https://github.com/Expensify/App/blob/821edc51788b616e323cae0dc00f4b0aa7246766/src/pages/workspace/WorkspaceNewRoomPage.tsx#L161

```
            } else if (values.roomName === '#announce') {
                if (!ValidationUtils.hasAnnounceRoom(reports, values.policyID ?? '-1')) {
                    ErrorUtils.addErrorMessage(errors, 'roomName', 'This name is reserved for future use. Please choose a different name');
                } else {
                    ErrorUtils.addErrorMessage(errors, 'roomName', translate('newRoomPage.roomAlreadyExistsError'));
                }
            } else if (ValidationUtils.isReservedRoomName(values.roomName)) {
```

### What alternative solutions did you explore? (Optional)

**Comment by mkzie2** [comment]
We only create a workspace when 3 or more participants are added. So, should we update the error message for creating a new room? When a user tries to create a room named 'announce,' my suggestion is:

1. If the policyAnnounce room has not been created yet, should show "This name is reserved for future use. Please choose a different name" - This message informs the user that the room doesn't exist yet but could be created in the future.

2. If the policyAnnounce room has been, should show "A room with this name already exists" or "Announce is a default room on all workspaces. Please choose another name." - This clarifies that the room is already created."


What do you think? @Expensify/design @sonialiap

**Comment by sonialiap** [comment]
@IuliiaHerets can you please test this and confirm the URL of the new announce room you seem to be creating with the mention? I my test the mention links to the workspace announce room, so no new room is being created. I can't repo

https://github.com/user-attachments/assets/6ab89dd9-a12d-4315-840f-bb35b86b8091

**Comment by MelvinBot** [comment]
This has been labelled "Needs Reproduction". Follow the steps here: https://stackoverflowteams.com/c/expensify/questions/16989

**Comment by IuliiaHerets** [comment]
@sonialiap QA team can repro this issue. The link from both rooms are different. Also, there are two #announce in LHN 

https://github.com/user-attachments/assets/bf4983d6-9e10-4e9f-9a77-3ce7dad4c538

**Comment by sonialiap** [comment]
Thanks for the replication and showing the URL for the confirmation. Does indeed look like we're duplicating chat rooms which is a bug

**Comment by sonialiap** [comment]
@Expensify/design problem: users can create rooms with the name `#announce`. Two solution ideas:

### one
https://github.com/Expensify/App/issues/50173#issuecomment-2392633411

> We only create a workspace when 3 or more participants are added. So, should we update the error message for creating a new room? When a user tries to create a room named 'announce,' my suggestion is:
> 1. If the policyAnnounce room has not been created yet, show "This name is reserved for future use. Please choose a different name" - This message informs the user that the room doesn't exist yet but could be created in the future.
> 2. If the policyAnnounce room has been, should show "A room with this name already exists" or "Announce is a default room on all workspaces. Please choose another name." - This clarifies that the room is already created.

### two
Or do we want to have a single message that says something like, `The name #announce is a default room title and cannot be used in a new name, please select a new name.` regardless of whether #announce already exists or if it hasn't been created yet?

**Comment by dannymcclain** [comment]
>Or do we want to have a single message that says something like, The name #announce is a default room title and cannot be used in a new name, please select a new name. regardless of whether #announce already exists or if it hasn't been created yet?

I think this idea is ok... But also, if the user _wants_ an `#announce` room, I feel like we should just give it to them at that point. From a technical standpoint I'm not sure how this could work, but this situation is feeling a little silly to me:

User: "Hey I need an `#announce` room!"
Expensify: "Nope. Can't do it. That's a default workspace room."
User: "Ok. I want it. For my workspace."
Expensify: "Nope. We'll probably automatically create that room later when you have more members."
User: "...Ok. Can I have it now instead?"
Expensify: "Nope 🙃"

Curious what the rest of the @Expensify/design team thinks. But I guess what I'm wondering is if there's any way to just go ahead and actually create the announce room from the get go when a workspace is created (and hide it or something? Unless someone _tries_ to find it?) or to just go ahead and create it if it hasn't been created yet.

**Comment by shawnborton** [comment]
I think I am more in the camp of keeping #announce as a protected room name since we have some logic that determines if/when the announce room gets created depending on how many members your workspace has. So I basically agree with Sonia's [two](https://github.com/Expensify/App/issues/50173#issuecomment-2410975110) above.

**Comment by mkzie2** [comment]
[Proposal updated](https://github.com/Expensify/App/issues/50173#issuecomment-2392625689)

**Comment by dubielzyk-expensify** [comment]
Yeah I tend to agree with @shawnborton here since they might eventually end up with an announce room which'll make things complicated.

**Comment by shawnborton** [comment]
I think we can use different error copy, something like:

"Default room names cannot be used as names for custom rooms"

As for the error, I think we'd use the standard form error patterns we have where the error shows above the submit button as well as below the field:

![CleanShot 2024-10-15 at 11 41 57@2x](https://github.com/user-attachments/assets/295c3656-a573-4166-9fe8-8956d686a065)

**Comment by shawnborton** [comment]
I think we'd just do nothing here and just display #announce as plain text if the room doesn't exist yet.

**Comment by mkzie2** [comment]
@shawnborton Currently, we already have the error message for the ```announce``` and ```admins``` cases:


![Screenshot 2024-10-15 at 18 53 05](https://github.com/user-attachments/assets/4eaaf08a-7e6d-43f4-bec5-0e88d5f8b796)
![Screenshot 2024-10-15 at 18 53 27](https://github.com/user-attachments/assets/08c844e4-acf9-424d-bf13-9f4053d3e8ec)

So you want to update the error message for ```announce``` case to "Default room names cannot be used as names for custom rooms", right?

**Comment by shawnborton** [comment]
Oh sorry, I didn't mean to distract - I guess this is my fault for not reading the entire issue history here! 

So I think I would just do nothing with the existing "Start a chat" flows - we already have error messages there, it's already covered. 

Going back to why this issue was created in the first place - I think I would just not send any kind of whisper when you write #announce or #admins if those rooms don't exist yet.

**Comment by yuwenmemon** [comment]
> First, BE should not return actionable whisper message when user send #announce message.

@mkzie2 Is this all you need in terms of backend changes?

**Comment by mkzie2** [comment]
> Is this all you need in terms of backend changes?

Yes

**Comment by sonialiap** [comment]
@yuwenmemon does this proposal look good? If yes, could you assign the contributor to the issue? https://github.com/Expensify/App/issues/50173#issuecomment-2428407731

**Comment by mkzie2** [comment]
@sonialiap Could you please add `HOLD` to the issue title to indicate that we're waiting for BE fix to be in place? Thanks!

**Comment by yuwenmemon** [comment]
Oops sorry - haven't been seeing this one as I haven't been the starred assignee. Making myself starred.

**Comment by yuwenmemon** [comment]
Haha I should have a PR for this imminently.

**Comment by yuwenmemon** [comment]
Still waiting on the backend to be deployed then will switch the star back to you @allgandalf

**Comment by yuwenmemon** [comment]
@mkzie2 Can you please give it a spin?

**Comment by mkzie2** [comment]
I tested the app after BE changes, and there is only one bug we need to fix from FE side: When we send the "#announce" message offline, it is highlighted until we are back online:

<img width="1130" alt="Screenshot 2024-11-24 at 16 17 58" src="https://github.com/user-attachments/assets/8f317033-a366-4599-8241-d8a8f4fb72da">


Do you agree it is bug @yuwenmemon @allgandalf?

**Comment by mkzie2** [comment]
Since [this behavior](https://github.com/Expensify/App/issues/50173#issuecomment-2495886399) occurs throughout the app and not just when sending #announce, I believe we need confirmation from the internal team on whether we should address the bug mentioned in [this comment](https://github.com/Expensify/App/issues/50173#issuecomment-2495886399) to ensure no effort is wasted. cc @yuwenmemon

**Comment by yuwenmemon** [comment]
Yep, that's a FE issue - let's fix it!

**Comment by yuwenmemon** [comment]
@mkzie2 do you want to tackle it?

**Comment by mkzie2** [comment]
To address the issue https://github.com/Expensify/App/issues/50173#issuecomment-2495886399, we should add ```exactlyMatch: true``` to:
https://github.com/Expensify/App/blob/f9ec4fa032a9a9d3edc0443f441246d81ed0e149/src/pages/home/report/ReportActionItem.tsx#L415

So it will be:

```
   const mentionReportContextValue = useMemo(() => ({currentReportID: report?.reportID ?? '-1', exactlyMatch: true}), [report?.reportID]);
```

What do you think @allgandalf ?

**Comment by yuwenmemon** [comment]
Not overdue, the proposal just came in. What do you think @allgandalf?

**Comment by sonialiap** [comment]
I'm OOO Dec 16-20, but I don't think this will need a BZ before I'm back so not adding a buddy

**Comment by mkzie2** [comment]
This change ensures that the message "#room-name" is only highlighted if "room-name" matches an existing room. As mentioned earlier, this will apply to all room names, not just #announce.

**Comment by mkzie2** [comment]
1. If we type: #test-room (Non existing room)

2. Then, It will not be highlighted. 

3. Then, BE returns the actionable message "Heads up, #test-room doesn't exist yet. Do you want to create it?"

4. If the user chooses "Yes", it will not be highlighted unless we get a response from the BE

**Comment by mkzie2** [comment]
Here is the test result:

https://github.com/user-attachments/assets/86097094-d535-48df-a6c4-d8f1cff46e62

**Comment by yuwenmemon** [comment]
Yeah, I agree, @allgandalf - so when you click "No," then we remove the highlight.

**Comment by mkzie2** [comment]
Yeah, I agree, @allgandalf - so when you click "No," then we remove the highlight.

We apply the highlight style to the message `#Test-room` to indicate that "Test-room" is an existing room.

In the above behavior, the highlight style is only removed if the user selects "No." This means the room name remains highlighted even if it is not an existing room. Also, that behavior conflicts with the current implementation in the latest main branch, where the highlight style is applied to the message `#Test-room` only when "Test-room" is an existing room.

@yuwenmemon Are you sure you want to implement the new behavior "when you click "No," then we remove the highlight"?

**Comment by yuwenmemon** [comment]
Yeah.

**Comment by mkzie2** [comment]
Will open in several hours.

**Comment by mkzie2** [comment]
**Here is the new behavior**:

1. User type: `#test-room` (Non existing room).

2. The `#test-room` message should be highlighted.

3. Then, BE returns the actionable message "Heads up, #test-room doesn't exist yet. Do you want to create it?"

4. The `#test-room` message is still highlighted.

5. If the user chooses "No", we remove the highlight.
    If the user chooses "Yes," we keep the highlight.


**Problem**: During implementation, an issue arises: From the context of `#test-room`, it is impossible to identify its corresponding actionable message. As a result, we cannot determine whether the actionable message for `#test-room` has been resolved with "Yes" or "No.". Consequently, we cannot decide whether to keep or remove the highlight style.

The core issue is the lack of a relationship between the `#test-room` message and its corresponding actionable message. Without such a link, it is impossible to map one to the other.


**Suggestion**: To address this, we can introduce a new property in the report action, such as `linkedActionableActionID`, within the `#test-room` message. This property would store the ID of the actionable message associated with the `#test-room` message, allowing us to retrieve the actionable message from the `#test-room` message.

What are your thoughts on the described problem and the proposed solution? Do you believe the addition of a `linkedActionableActionID` property is a viable and effective approach to solving this issue? This needs to be handled on both BE and FE sides. @yuwenmemon @allgandalf

**Comment by shawnborton** [comment]
Let us know what questions you have, though it seems like the linked comment above is more of a code-related question than it is anything related to the UI/UX.

**Comment by dannymcclain** [comment]
>Personally, I believe the second approach makes more sense, as it only highlights #test-room if the room actually exists. Let’s see what the design team thinks @shawnborton.

Yeah I tend to agree. I think it makes sense to only highlight things we know exist. cc @Expensify/design

**Comment by shawnborton** [comment]
Totally, that makes sense to me too 👍

**Comment by dubielzyk-expensify** [comment]
Agree with the designers here as well 👍 Seems clearer to me

**Comment by sonialiap** [comment]
Thanks @allgandalf for the checklist! Payment made
